lib/lpc55-rot-startup/src/images.rs

```diff
@@ -56,129 +38,439 @@ extern "C" {
-                // the build.
-                let page = unsafe {
+        // TODO: Find the caboose if present and reduce end_offset.
+
+        // SAFETY: Link time constants known to be safe.
+        let mut begin_offset =
+            unsafe { core::ptr::addr_of!(__vector_size) as u32 };
+        if self.is_imageheader_present(false).is_ok() {
+            // SAFETY: Link time constants known to be safe.
+            begin_offset += core::mem::size_of::<ImageHeader>() as u32;
+        }
+        if let Some(end) = self.run.end.checked_add(end_offset) {
+            // SAFETY: Link time constants known to be safe.
+            Ok(self.run.start + begin_offset..end)
+        } else {
+            Err(ImageError::InvalidLength)
+        }
+    }
+
+    // Assuming a well-formed image, get the result of the ROM signature check
+    // routine.
+    // If configured to do so, CRC type images can also be checked.
+    fn check_signature(&self) -> Result<(), ImageError> {
+        // The following code is adapted to fit here from bootleby.
+
+        // The image type is at offset 0x24
+        //   - 0x0000 Normal image for unsecure boot
+        //   - 0x0001 Plain signed Image
+        //   - 0x0002 Plain CRC Image, CRC at offset 0x28
+        //   - 0x0004 Plain signed XIP Image
+        //   - 0x0005 Plain CRC XIP Image, CRC at offset 0x28
+        //   - 0x8001 Signed plain Image with KeyStore Included
+        match self.get_image_type() {
+            4 => (),
+            #[cfg(feature = "allow-unsigned-images")]
+            5 => {
+                // CRC checksum used by the ROM. It'd be great if the ROM would
+                // check this for us, wouldn't it?
+                //
+                // It won't though; there's no entry point for it.
+
+                // We have already verified that image_length seems plausible.
+
+                let crc = Crc::<u32>::new(&CRC_32_MPEG_2);
+                let mut digest = crc.digest();
+
+                let data = unsafe {
                     core::slice::from_raw_parts(
-                        start as *const u8,
-                        BYTES_PER_FLASH_PAGE,
+                        flash.start as *const u8,
+                        image_length,
                     )
                 };
-                hash.update(page);
+
+                // The CRC calculation simply _skips_ the CRC word, as opposed
+                // to (say) including it with an assumed value of 0. So to CRC
+                // our header we have to jump through...a small hoop.
+                digest.update(&data[..0x28]);
+                digest.update(&data[0x28 + 4..]);
+                let computed_crc = digest.finalize();
+                // The CRC format (ab)uses the `type_specific_header` field to
+                // store the CRC.
+                if computed_crc != self.get_type_specific_header() {
+                    return Err(ImageError::Crc);
+                }
```

Comment:
bootleby is intended to be very generic. I don't think we want that here and I think we should skip checking for CRC images.

Reply:
Removed CRC support.